### PR TITLE
Vest a portion of block reward immediately (FIP-0004)

### PR DESCRIPTION
**Pull request by ZenGround0**
- [x] Code should be correct
- [x] Unrelated test failures caused by those tests using applyRewards as a test state setup helper are now fixed.
- [x] Reward tests updated to test new behavior
- [x] Reward test exercising rewards in both versions to catch any unexpected balance table assumptions about not having consistent vesting specs applied.
- [x] Land #1251 and rebase on master
- [ ] Update PR title when FIP is published if the number changes

### CODE REVIEW

File: actors/builtin/miner/miner_actor.go
Comment (anorth) [inline actors/builtin/miner/miner_actor.go:1510]:
```suggestion
			rt.Abortf(exitcode.ErrInsufficientFunds, "insufficient funds to lock, available: %v, requested: %v", unlockedBalance, rewardToLock)
```

File: actors/builtin/miner/monies.go
Comment (anorth) [inline actors/builtin/miner/monies.go:178]: My preferred pattern here is to pass the network version as a parameter here rather than switch on it at the caller. In general, this would reduce the number of places that need to get the version comparison right, and statically find all the relevant places. (because this is introducing a new function, previously identity, we don't get the static win here).

File: actors/builtin/miner/miner_test.go
Comment (anorth) [inline actors/builtin/miner/miner_test.go:1309]: I think in general all unit tests should target the latest network version. If we have a version switch, it will generally be because the prior version is already live in the network, and so we're working to test that the new version will behave as expected. E.g. maybe there's some unexpected interaction between the change guarded by this version and other things that are being tested here.

Which means generally all these tests that are not specifically about the reward vesting should require no diff (because the runtime uses VersionMax). If they do require a diff, that diff will specifically highlight the behaviour that's changing.

Is there a diff here to pledge expectations or something? I think we should update expectations to match (hopefully we can do it just once in the deadlineCron or something).

File: actors/builtin/miner/miner_test.go
Comment (anorth) [inline actors/builtin/miner/miner_test.go:3963]: Here is where we can have a test for both the old and new versions. Grandfather the old one by setting the network version to the prior one, and add a new one.

File: actors/builtin/miner/miner_test.go
Comment (ZenGround0) [reply actors/builtin/miner/miner_test.go:1309]: > I think in general all unit tests should target the latest network version.

agree

> Which means generally all these tests that are not specifically about the reward vesting should require no diff

agree on the *should*

> Is there a diff here to pledge expectations or something?

Yes.  The problem is that we use applyRewards as a testing setup function in a bunch of places in a fragile way, not as part of actually testing "what happens if the miners gets a reward message and then x and y happen".  There is a better way to do it as we've known for a while, however designing it and then this refactor seems non-trivial to me.  I didn't want fixing up our testing problems to block this change so I used the ugly approach you see in the diff.  The benefit of this approach is that it makes it really obvious that there is a problem with how we setup tests, because as you point out we shouldn't be using v5 in tests.

Comment (anorth) [reply actors/builtin/miner/miner_test.go:1309]: I have addressed this in #1251 and will rebase this PR upon that one, removing the noise.

File: actors/builtin/miner/monies.go
Comment (anorth) [reply actors/builtin/miner/monies.go:178]: Done.

File: actors/builtin/miner/miner_test.go
Comment (anorth) [reply actors/builtin/miner/miner_test.go:3909]: Actually now that I've reverted the change to initial delay, there's not really anything to grandfather here. This is just an improvement on the old test.

**Comment by anorth** [comment]
@ZenGround0 @acruikshank  I have applied my own suggestions. Please carefully review the commits I added here, then I think it's good to go (after #1251).

**Review by anorth — APPROVED** [review]

File: actors/builtin/miner/miner_test.go
Comment (ZenGround0) [inline actors/builtin/miner/miner_test.go:5106]: Nice demonstration of some of the benefits of pushing the network version down

**Review by acruikshank — APPROVED** [review]
LGTM

**Pull request merged**
